packages/kit/src/runtime/client/client.js

```diff
@@ -1173,6 +1179,77 @@ export function create_client({ target, base }) {
 		});
 	}
 
+	function setup_preload() {
+		/** @type {NodeJS.Timeout} */
+		let mousemove_timeout;
+
+		target.addEventListener('mousemove', (event) => {
+			const target = /** @type {Element} */ (event.target);
+
+			clearTimeout(mousemove_timeout);
+			mousemove_timeout = setTimeout(() => {
+				preload(target, 2);
+			}, 20);
+		});
+
+		/** @param {Event} event */
+		function tap(event) {
+			preload(/** @type {Element} */ (event.composedPath()[0]), 1);
+		}
+
+		target.addEventListener('mousedown', tap);
+		target.addEventListener('touchstart', tap, { passive: true });
+
+		const observer = new IntersectionObserver(
+			(entries) => {
+				for (const entry of entries) {
+					if (entry.isIntersecting) {
+						preload_code(new URL(/** @type {HTMLAnchorElement} */ (entry.target).href).pathname);
+						observer.unobserve(entry.target);
+					}
+				}
+			},
+			{ threshold: 0 }
+		);
+
+		/**
+		 * @param {Element} element
+		 * @param {number} priority
+		 */
+		function preload(element, priority) {
+			const { url, options, external } = find_anchor(element, base);
+
+			if (!external) {
+				if (priority <= options.preload_data) {
+					preload_data(/** @type {URL} */ (url));
+				} else if (priority <= options.preload_code) {
+					preload_code(/** @type {URL} */ (url).pathname);
+				}
+			}
+		}
+
+		function after_navigate() {
+			observer.disconnect();
+
+			for (const a of target.querySelectorAll('a')) {
+				const { url, external, options } = find_anchor(a, base);
+
+				if (external) continue;
+
+				if (options.preload_code === PRELOAD_PRIORITIES.viewport) {
+					observer.observe(a);
+				}
+
+				if (options.preload_code === PRELOAD_PRIORITIES.page) {
+					preload_code(/** @type {URL} */ (url).pathname);
+				}
+			}
+		}
+
+		callbacks.after_navigate.push(after_navigate);
+		after_navigate();
```

Comment:
I'm wondering if this is needed.
Callbacks in `callbacks.after_navigate` are called after hydration anyway, which means this is executed twice for the initial page.

Reply:
It's needed. This setup happens _after_ the page has hydrated, meaning the callback won't be called until the next navigation